[CO-417] Add more friendly 415 Unsupported Media Type error

paweljakubas opened the pull request:
## Description

Implement more friendly 415 Unsupported Media Type error.

**BEFORE**
```
$ curl -X POST -vk -H "Content-Type: patate" http://localhost:8090/api/v1/wallets 

> POST /api/v1/wallets HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.60.0
> Accept: */*
> Content-Type: patate
> 
< HTTP/1.1 415 Unsupported Media Type
< Transfer-Encoding: chunked
< Date: Fri, 05 Oct 2018 08:57:20 GMT
< Server: Warp/3.2.22
< 
* Connection #0 to host localhost left intact
```

**AFTER**
```
$ curl -X POST -vk -H "Content-Type: patate" http://localhost:8090/api/v1/wallets

> POST /api/v1/wallets HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.60.0
> Accept: */*
> Content-Type: patate
> 
< HTTP/1.1 415 Unsupported Media Type
< Transfer-Encoding: chunked
< Date: Mon, 08 Oct 2018 13:14:42 GMT
< Server: Warp/3.2.22
< Content-Type: application/json
< 
* Connection #0 to host localhost left intact
{"status":"error","diagnostic":{"mimeContentTypeError":"The API expects the Content-Type's main MIME-type to be 'application/json'"},"message":"UnsupportedMimeTypePresent"}
```

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CO-417

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
1. Run cluster and 
```
curl -X POST -vk -H "Content-Type: patate" http://localhost:8090/api/v1/wallets
```

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

paweljakubas commented:
`continuous-integration/appveyor/pr — AppVeyor build failed` because of timeout

paweljakubas commented:
Made changes as requested:
1. Introduced `UnsupportedMimeTypeError`
2. Encapsulated the stuff inside middleware
3. Added golden test, swagger entry and corresponding `aesonRoundtripProp`

Code rebased, and squashed

paweljakubas commented:
Both AppVeyor failures are due to timeouts

KtorZ commented on wallet-new/src/Cardano/Wallet/Server/Middlewares.hs:
```diff
@@ -29,6 +36,21 @@ import           Pos.Launcher.Configuration (ThrottleSettings (..))
+    | otherwise = r
+
+unsupportedMimeTypeMiddleware :: Middleware
+unsupportedMimeTypeMiddleware = modifyResponse responseModifier
```
Since `responseModifier` is solely related to the above function, it would make sense to have it under a `where` clause. It doesn't make sense to use `responseModifier` as it is in any other context :+1:

KtorZ commented:
May you add an entry in the CHANGELOG, explaining that we now have more developer-friendly errors for wrong content-type ?

Incidentally, that is a breaking change :upside_down_face:  ...

KtorZ commented on wallet-new/src/Cardano/Wallet/API/Response.hs:
```diff
@@ -269,3 +270,36 @@ instance HasDiagnostic JSONValidationError where
+
+instance Buildable UnsupportedMimeTypeError where
+    build _ =
+        bprint "The API expects the Content-Type's main MIME-type to be 'application/json'"
```
I think this should display what get passed as `Text` to the `UnsupportedMimeTypePresent` constructor ? Don't you think?

KtorZ commented on wallet-new/test/WalletNewJson.hs:
```diff
@@ -165,3 +166,12 @@ golden_JSONValidationError_JSONValidationFailed =
+golden_UnsupportedMimeTypeError_UnsupportedMimeTypePresent =
+    goldenTestJSON
+        (UnsupportedMimeTypePresent "test")
+            "test/golden/UnsupportedMimeTypeError_UnsupportedMimeTypePresent"
```
:heart:

paweljakubas commented:
Changes:
1. Added entry to CHANGELOG.md
2. Refactored `instance Buildable UnsupportedMimeTypeError`
3.  Move `responseModifier` into `where`

Commits squashed, was up to date against develop
